File: baekjoon/python/7569/7569.py
from collections import deque
import sys
sys.stdin = open('7569_input.txt', 'r')


m, n, h = map(int, input().split())
graph = []
for _ in range(h):
    # z축이 맨 앞
    temp = [list(map(int, input().split())) for _ in range(n)]
    graph.append(temp)
dx = [0, 0, 1, -1, 0, 0]
dy = [1, -1, 0, 0, 0, 0]
dz = [0, 0, 0, 0, 1, -1]


def gr(graph):
    for i in graph:
        for j in i:
            print(j)
        print()
    print()


def bfs(start):
    # 익은 토마토에서 시작
    q = deque()
    for riped in start:
        q.append(riped)
    # print(q/)
    while q:
        # gr(graph)
        z, x, y = q.popleft()
        for i in range(6):
            nx = x + dx[i]
            ny = y + dy[i]
            nz = z + dz[i]
            # 벽이 아닐때만 진행
            if 0 <= nz < h and 0 <= nx < n and 0 <= ny < m:
                # 안익은 토마토일때
                if graph[nz][nx][ny] == 0:
                    # 익는 날짜 추가
                    graph[nz][nx][ny] = graph[z][x][y] + 1
                    # 내일 익는 토마토 큐에 추가
                    q.append((nz, nx, ny))


# 익은 토마토 위치 확인
pos = []
for k in range(h):
    for i in range(n):
        for j in range(m):
            if graph[k][i][j] == 1:
                pos.append((k, i, j))
# print(pos)
bfs(pos)

fail = False

# gr(graph)

MAX = []
for lst in graph:
    for elem in lst:
        if 0 in elem:
            fail = True
            break
    MAX.append(max(map(max, lst)))

if fail:
    print(-1)
else:
    print(max(MAX)-1)

File: baekjoon/cpp/7569/7569.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <tuple>

using namespace std;

int n, m, h;

int graph[101][101][101];
bool visited[101][101][101] = {
    false,
};

void bfs(queue<tuple<int, int, int>> &q)
{
    int dx[6] = {1, -1, 0, 0, 0, 0};
    int dy[6] = {0, 0, 1, -1, 0, 0};
    int dz[6] = {0, 0, 0, 0, -1, 1};
    int nx, ny, nz;
    while (!q.empty())
    {
        int x, y, z;
        tie(x, y, z) = q.front();
        q.pop();
        for (int i = 0; i < 6; i++)
        {
            nx = x + dx[i];
            ny = y + dy[i];
            nz = z + dz[i];
            if (0 <= nx && nx < n && 0 <= ny && ny < m && 0 <= nz && nz < h &&
                !visited[nx][ny][nz] && graph[nx][ny][nz] != -1)
            {
                if (!graph[nx][ny][nz])
                {
                    visited[nx][ny][nz] = true;
                    graph[nx][ny][nz] = graph[x][y][z] + 1;
                    q.push({nx, ny, nz});
                }
            }
        }
    }
}

int main()
{

    std::ios_base::sync_with_stdio(false);
    cin.tie(NULL);
    freopen("7569_input.txt", "r", stdin);

    cin >> m >> n >> h;
    int tmp;
    queue<tuple<int, int, int>> tom_q;
    for (int k = 0; k < h; k++)
    {
        for (int i = 0; i < n; i++)
        {
            for (int j = 0; j < m; j++)
            {

                cin >> tmp;
                graph[i][j][k] = tmp;
                if (tmp == 1)
                {
                    tom_q.push({i, j, k});
                    visited[i][j][k] = true;
                }
            }
        }
    }

    // while (!tom_q.empty())
    // {
    //     int x, y, z;
    //     tie(x, y, z) = tom_q.front();
    //     cout << x << "," << y << "," << z << endl;
    //     tom_q.pop();
    // }
    // for (int k = 0; k < h; k++)
    // {
    //     for (int i = 0; i < n; i++)
    //     {
    //         for (int j = 0; j < m; j++)
    //         {

    //             cout << graph[i][j][k] << " ";
    //         }
    //         cout << endl;
    //     }
    //     cout << endl;
    // }
    // cout << endl;

    bfs(tom_q);

    // for (int k = 0; k < h; k++)
    // {
    //     for (int i = 0; i < n; i++)
    //     {
    //         for (int j = 0; j < m; j++)
    //         {

    //             cout << graph[i][j][k] << " ";
    //         }
    //         cout << endl;
    //     }
    //     cout << endl;
    // }

    int max = -1;

    for (int k = 0; k < h; k++)
    {
        for (int i = 0; i < n; i++)
        {
            for (int j = 0; j < m; j++)
            {

                if (graph[i][j][k] == 0)
                {
                    cout << -1 << endl;
                    return 0;
                }
                else
                {
                    if (graph[i][j][k] > max)
                        max = graph[i][j][k];
                }

                // for (const auto &elem : graph[i])
                // {
                //     cout << elem << " ";
                // }
                // cout << endl;
            }
        }
    }
    cout << max - 1 << endl;
    return 0;
}
